packages/SwingSet/src/kernel/kernel.js

```diff
@@ -705,6 +705,10 @@ export default function buildKernel(
       }
       commitCrank();
       kernelKeeper.incrementCrankNumber();
+      if (snapStore) {
+        // eslint-disable-next-line no-use-before-define
+        vatWarehouse.pruneSnapshots(snapStore, kernelKeeper.getCrankNumber());
+      }
```

Comment:
unfortunately this is too early, because of the following scenario:

* host commits block 1
* host starts processing block 2
  * kernel writes snapshot4 for vatA, causing the previous snapshot3 to become unreferenced
  * 20 more cranks happen, enoguh to trip `snapshotPruneInterval`
  * crank 21 finishes, `commitCrank` writes it to the crank buffer
    * note that all 21 cranks are still sitting in the block buffer, since the host has not yet gotten control back to write out block 2
  * kernel calls `pruneSnapshots`, snapshot3 is deleted
  * host process dies
* host process restarts, tries to process block 2 again
  * DB shows vatA using snapshot3
  * vat-warehouse attempts to load vatA from snapshot3, fails because the file is deleted

We need to defer the deletion until after the host has committed the block buffer. We'll need to figure out the best place to do this.. it's not dissimilar to the problems the `streamStore` has/had w.r.t. external files. If we could store snapshots in the same DB (within the same atomicity domain) as the rest of the kernel state, that'd fix this problem, but 1: the snapshots are larger than we'd like to put there, and 2: they need to be available on disk for the `xsnap` process to see, so there's still a "when does it get added/removed in the DB" question to answer.

Reply:
Storing the transcript in the stream store had something of the opposite problem, namely that it would write entries reflecting a future that didn't happen and needed to be rolled back, pruning the timeline at the front.  Since the last thing written to the DB is the endpoint, it was safe to leave the record of that future out there and just let the new timeline overwrite it.  Consequently I'm not sure the stream store experience has much to tell us.  This is more like pulling up the carpet behind you.

Reply:
@dckc: @FUDCo and I walked through this today, and figured that the best solution is something like:
* buffer the deletion inside the `snapstore`: when the kernel/vat-wardhouse checks the refcounts and sees that the snapshot is eligible for deletion, it calls `snapstore.delete()`, but that merely puts an entry in a `deletionSet` Set
* when creating a snapshot (just after we determine the name of the snapshot, by hashing its contents), check to see if it's already in `deletionSet`, and if so, remove it
* consolidate `snapstore` with the rest of `swingstore`
* when the swingstore's `commit()` is called (by the host app), swingstore leans over to snapstore and says "hey, now would be a safe time to delete things"
* snapstore then walks through the `deletionSet` and deletes the files therein

This protects against the case where the kernel decides to delete a snapshot, but is then interrupted before the kvstore updates which stop using it are committed. The delete always happens after that kvstore commit, during the "safe deletion window" that begins after the kvstore commit and ends when a new snapshot is created and committed to the kvstore.

This does leave a small window for leaks: if the process is interrupted after the kvstore commit, but before the file deletion: the kernel will have removed the 0 refcount kvstore entry at the same time as it enqueues the deletion, and that kvstore removal has been committed, but the deletion didn't actually happen, and when the process is restarted, the kernel will no longer remember what needs to be deleted. The core problem is that we have two atomicity domains.

This relates to the #3431 refactoring we were talking about. It might possible to reduce the size of the change by doing something like `makeLMDBSwingStore(basedir, snapstore)`, so one can call the `commit` of the other at the right time. Or by adding a `commit` to the snapstore, and then changing the host applications to commit both (always kvstore first, then snapstore), but it's probably better to give applications fewer opportunities to get it wrong.

Reply:
> @dckc: @FUDCo and I walked through this today, and figured that the best solution is something like:

Yes, this was pretty straightforward...

> * buffer the deletion inside the `snapstore`: ... `snapstore.delete()`

I called it `prepareToDelete`

> * consolidate `snapstore` with the rest of `swingstore`

This involved some coin-tosses. LMK if looks OK.

> This relates to the #3431 refactoring we were talking about. It might possible to reduce the size of the change by doing something like `makeLMDBSwingStore(basedir, snapstore)` ...

Since `makeLMDBSwingStore` already creates the `streamStore`, I figured it might as well create the `snapStore` too.